src/EditorFeatures/CSharp/CompleteStatement/CompleteStatementCommandHandler.cs

```diff
@@ -0,0 +1,456 @@
+                    return ((StructDeclarationSyntax)currentNode).SemicolonToken.IsMissing;
+                case SyntaxKind.InterfaceDeclaration:
+                    return ((InterfaceDeclarationSyntax)currentNode).SemicolonToken.IsMissing;
+                case SyntaxKind.EnumDeclaration:
+                    return ((EnumDeclarationSyntax)currentNode).SemicolonToken.IsMissing;
+                case SyntaxKind.DelegateDeclaration:
+                    return ((DelegateDeclarationSyntax)currentNode).SemicolonToken.IsMissing;
+                case SyntaxKind.EventFieldDeclaration:
+                    return ((EventFieldDeclarationSyntax)currentNode).SemicolonToken.IsMissing;
+                case SyntaxKind.MethodDeclaration:
+                    return ((MethodDeclarationSyntax)currentNode).SemicolonToken.IsMissing;
+                case SyntaxKind.OperatorDeclaration:
+                    return ((OperatorDeclarationSyntax)currentNode).SemicolonToken.IsMissing;
+                case SyntaxKind.ConversionOperatorDeclaration:
+                    return ((ConversionOperatorDeclarationSyntax)currentNode).SemicolonToken.IsMissing;
+                case SyntaxKind.ConstructorDeclaration:
+                    return ((ConstructorDeclarationSyntax)currentNode).SemicolonToken.IsMissing;
+                case SyntaxKind.BaseConstructorInitializer:
+                case SyntaxKind.ThisConstructorInitializer:
+                    return ((ConstructorDeclarationSyntax)currentNode).SemicolonToken.IsMissing;
+                case SyntaxKind.DestructorDeclaration:
+                    return ((DestructorDeclarationSyntax)currentNode).SemicolonToken.IsMissing;
+                case SyntaxKind.PropertyDeclaration:
+                    return ((PropertyDeclarationSyntax)currentNode).SemicolonToken.IsMissing;
+                case SyntaxKind.IndexerDeclaration:
+                    return ((IndexerDeclarationSyntax)currentNode).SemicolonToken.IsMissing;
+                case SyntaxKind.AddAccessorDeclaration:
+                    return ((AccessorDeclarationSyntax)currentNode).SemicolonToken.IsMissing;
+                default:
+                    return false;
+            }
+        }
+
+        private static bool StatementClosingDelimiterExists(SyntaxNode currentNode, ref TextSpan lastDelimiterSpan)
+        {
+            switch (currentNode.Kind())
+            {
+                case SyntaxKind.DoStatement:
+                    var dostatement = (DoStatementSyntax)currentNode;
+                    lastDelimiterSpan = dostatement.CloseParenToken.Span;
+                    return !dostatement.CloseParenToken.IsMissing;
+
+                default:
+                    // Statement I'm not handling yet so shouldn't proceed with statement completion
+                    return false;
+            }
+        }
+
+        private static bool ClosingDelimiterExistsIfNeeded(SyntaxNode currentNode, ref TextSpan lastDelimiterSpan)
```

Comment:
stopping at this point.  most of these algorithms need explanations/examples to help explain what's going on.